streams/src/test/java/org/apache/kafka/streams/integration/IQv2StoreIntegrationTest.java

```diff
@@ -0,0 +1,557 @@
+            IntegrationTestUtils.iqv2WaitForPartitionsOrGlobal(kafkaStreams, request, partitions);
+
+        makeAssertions(
+            partitions,
+            result,
+            queryResult -> {
+                final boolean failure = queryResult.isFailure();
+                if (failure) {
+                    assertThat(queryResult.toString(), failure, is(false));
+                }
+                assertThat(queryResult.isSuccess(), is(true));
+
+                // TODO: position not implemented
+                assertThat(queryResult.getPosition(), is(nullValue()));
+
+                assertThrows(IllegalArgumentException.class, queryResult::getFailureReason);
+                assertThrows(IllegalArgumentException.class, queryResult::getFailureMessage);
+
+                assertThat(queryResult.getResult(), is(true));
+
+                assertThat(queryResult.getExecutionInfo(), is(empty()));
+            });
+    }
+
+    public void shouldCollectExecutionInfo() {
+
+        final PingQuery query = new PingQuery();
+        final StateQueryRequest<Boolean> request =
+            inStore(STORE_NAME).withQuery(query).enableExecutionInfo();
+        final Set<Integer> partitions = mkSet(0, 1);
+
+        final StateQueryResult<Boolean> result =
+            IntegrationTestUtils.iqv2WaitForPartitionsOrGlobal(kafkaStreams, request, partitions);
+
+        makeAssertions(
+            partitions,
+            result,
+            queryResult -> assertThat(queryResult.getExecutionInfo(), not(empty()))
+        );
+    }
+
+    public void shouldCollectExecutionInfoUnderFailure() {
+
+        final UnknownQuery query = new UnknownQuery();
+        final StateQueryRequest<Void> request =
+            inStore(STORE_NAME).withQuery(query).enableExecutionInfo();
+        final Set<Integer> partitions = mkSet(0, 1);
+
+        final StateQueryResult<Void> result =
+            IntegrationTestUtils.iqv2WaitForPartitionsOrGlobal(kafkaStreams, request, partitions);
+
+        makeAssertions(
+            partitions,
+            result,
+            queryResult -> assertThat(queryResult.getExecutionInfo(), not(empty()))
+        );
+    }
+
+    private <R> void makeAssertions(
+        final Set<Integer> partitions,
+        final StateQueryResult<R> result,
+        final Consumer<QueryResult<R>> assertion) {
+
+        if (result.getGlobalResult() != null) {
+            assertion.accept(result.getGlobalResult());
+        } else {
+            assertThat(result.getPartitionResults().keySet(), is(partitions));
+            for (final Integer partition : partitions) {
+                assertion.accept(result.getPartitionResults().get(partition));
+            }
+        }
+    }
+
+    private static Properties streamsConfiguration(final boolean cache, final boolean log,
+        final String supplier) {
+        final String safeTestName =
+            IQv2StoreIntegrationTest.class.getName() + "-" + cache + "-" + log + "-" + supplier;
```

Comment:
Note: to prevent tests from polluting each other, it's important for each test to have a unique application name and state directory.